

Olá estudante!

Me chamo Rafael Meirelles e irei revisar o seu projeto hoje e em eventuais futuras submissões até que ele cumpra todos os requisitos para o aceite.
Conte comigo nessa jornada e não se preocupe se precisar ajustar alguns detalhes, é parte do processo e fundamental para que você exercite os conceitos que vem aprendendo e assim melhore a qualidade dos seus códigos e análises.

**Peço que mantenha e não altereos comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:


<div class="alert alert-danger">
<strong>Vermelho</strong>

Erro que precisa ser consertado, caso contrário, seu projeto não pode ser aceito

</div>

<div class="alert alert-warning">
<strong>Amarelo</strong>

Alerta de um erro não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise

</div>

<div class="alert alert-success">
<strong>Verde</strong>

Elogios
</div>

<div class="alert alert-info">
<strong>Comentário do estudante</strong>

Use uma caixa azul como essa para eventuais comentários que você gostaria de fazer para mim.
</div>






<div class="alert alert-danger">
<strong>Comentário do revisor v1</strong>

Estamos no caminho certo com essa versão, mas retocar os pontos indicados abaixo nos fará chegar em um resultado ainda mais incrível.
    
Conte comigo para ajudá-lo (a) em qualquer questão que possa vir a ter! Lembre-se de usar as caixas azuis para se comunicar comigo.
    
Até breve.
</div>


<div class="alert alert-danger">
<strong>Comentário do revisor v2</strong>

Avançamos bastante com os ajustes, mas ainda temos alguns pontos pendentes. Procure por 'v2' para encontrá-los com mais facilidade.

Estamos quase lá! Vamos juntos até o fim.
</div>



<div class="alert alert-danger">
<strong>Comentário do revisor v3</strong>

Avançamos bem nessa iteração! Muito bom!
    
Estamos quase lá! Foque por favor nos pontos indicados como v3 e chegaremos a aprovação.
    
Até breve!
</div>



## Objetivos do Estudo

Com o aumento do consumo de livros durante o período da pandemia, diversas startups passaram a explorar esse comportamento como oportunidade de negócio. O objetivo deste estudo é realizar uma análise exploratória e quantitativa sobre os dados de uma plataforma digital de leitura para:

- Entender o perfil do catálogo de livros disponível (volume, recência, qualidade);
- Identificar editoras e autores com maior influência ou engajamento;
- Avaliar o comportamento de avaliação dos usuários, incluindo padrões de uso intenso;
- Apoiar a tomada de decisão sobre o desenvolvimento de um novo produto digital voltado ao público leitor, com base em dados reais de consumo e preferências.

Todas as análises serão conduzidas diretamente sobre um banco de dados PostgreSQL utilizando consultas SQL otimizadas.


In [2]:
# Importar bibliotecas necessárias
import pandas as pd
from sqlalchemy import create_engine

# Configurações da conexão
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-final-project-db'
}

# Criar a string de conexão
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

# Criar engine de conexão com o banco
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Exemplo de consulta SQL:
# query = 'SELECT * FROM books LIMIT 5;'
# pd.read_sql(query, con=engine)


In [3]:
# Visualizar amostra da tabela books
query_books_amostra = 'SELECT * FROM books LIMIT 5;'
pd.read_sql(query_books_amostra, con=engine)


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# Consulta SQL para contar livros publicados após 01/01/2000
query_livros_2000 = '''
SELECT COUNT(*) AS livros_pos_2000
FROM books
WHERE publication_date > '2000-01-01';
'''
resultado_livros_2000 = pd.read_sql(query_livros_2000, con=engine)
resultado_livros_2000


livros_pos_2000
0              819

Livros lançados após 1º de janeiro de 2000

Resultado
Foram identificados **819 livros** publicados após 1º de janeiro de 2000.


<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

A questão está correta!

</div>


In [5]:
# Amostra da tabela ratings
query_ratings_amostra = 'SELECT * FROM ratings LIMIT 5;'
pd.read_sql(query_ratings_amostra, con=engine)


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
# Consulta SQL: total de avaliações e média de classificação por livro
query_avaliacoes_livros = '''
SELECT 
    b.title AS titulo_livro,
    COUNT(r.rating_id) AS total_avaliacoes,
    ROUND(AVG(r.rating), 2) AS media_classificacao
FROM books b
JOIN ratings r ON b.book_id = r.book_id
GROUP BY b.title
ORDER BY total_avaliacoes DESC;
'''
resultado_avaliacoes = pd.read_sql(query_avaliacoes_livros, con=engine)
resultado_avaliacoes.head()


titulo_livro  total_avaliacoes  \
0                            Twilight (Twilight  #1)               160   
1                The Hobbit  or There and Back Again                88   
2                             The Catcher in the Rye                86   
3               Angels & Demons (Robert Langdon  #1)                84   
4  Harry Potter and the Prisoner of Azkaban (Harr...                82   

   media_classificacao  
0                 3.66  
1                 4.13  
2                 3.83  
3                 3.68  
4                 4.41

Os cinco livros com maior número de avaliações foram:

1. **Twilight (Twilight #1)** — 160 avaliações, média **3.66**
2. **The Hobbit or There and Back Again** — 88 avaliações, média **4.13**
3. **The Catcher in the Rye** — 86 avaliações, média **3.83**
4. **Angels & Demons (Robert Langdon #1)** — 84 avaliações, média **3.68**
5. **Harry Potter and the Prisoner of Azkaban** — 82 avaliações, média **4.41**


<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

A questão está correta!

</div>


In [7]:
# Visualizar as primeiras linhas das tabelas
pd.read_sql('SELECT * FROM books LIMIT 5;', con=engine)
pd.read_sql('SELECT * FROM publishers LIMIT 5;', con=engine)


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
# Consulta SQL para identificar a editora com mais livros com mais de 50 páginas
query_editora_maior_producao = '''
SELECT 
    p.publisher AS editora,
    COUNT(b.book_id) AS total_livros
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY total_livros DESC
LIMIT 1;
'''
resultado_editora = pd.read_sql(query_editora_maior_producao, con=engine)
resultado_editora


editora  total_livros
0  Penguin Books            42

Resultado
A editora que mais publicou livros com mais de 50 páginas foi:

- **Penguin Books** — com **42 livros completos** no catálogo.




<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

A questão está correta!

</div>


In [9]:
# Visualizar amostras
pd.read_sql('SELECT * FROM books LIMIT 5;', con=engine)
pd.read_sql('SELECT * FROM authors LIMIT 5;', con=engine)
pd.read_sql('SELECT * FROM ratings LIMIT 5;', con=engine)


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
# Autor com maior média de avaliação, considerando apenas livros com pelo menos 50 avaliações
query_autor_top = '''
SELECT 
    a.author AS autor,
    ROUND(AVG(sub.avg_rating), 2) AS media_geral
FROM (
    SELECT 
        b.book_id,
        b.author_id,
        AVG(r.rating) AS avg_rating,
        COUNT(r.rating_id) AS total_avaliacoes
    FROM books b
    JOIN ratings r ON b.book_id = r.book_id
    GROUP BY b.book_id
    HAVING COUNT(r.rating_id) >= 50
) sub
JOIN authors a ON sub.author_id = a.author_id
GROUP BY a.author
ORDER BY media_geral DESC
LIMIT 1;
'''
resultado_autor = pd.read_sql(query_autor_top, con=engine)
resultado_autor


autor  media_geral
0  J.K. Rowling/Mary GrandPré         4.28

Resultado
O autor com a maior média de classificação entre livros com pelo menos 50 avaliações foi:

- **J.K. Rowling / Mary GrandPré** — com média **4.28**.

<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

A questão está correta!

</div>


In [11]:
# Visualizar amostra da tabela ratings
pd.read_sql('SELECT * FROM ratings LIMIT 5;', con=engine)


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [12]:
query_reviews_corrigida = '''
WITH active_users AS (
    SELECT
        username
    FROM ratings
    GROUP BY username
    HAVING COUNT(rating_id) > 50
)
SELECT
    ROUND(AVG(review_count), 2) AS avg_reviews_per_user
FROM (
    SELECT
        username,
        COUNT(review_id) AS review_count
    FROM reviews
    WHERE username IN (SELECT username FROM active_users)
    GROUP BY username
) AS user_reviews_count;
'''

# Executar consulta
resultado_reviews_corrigida = pd.read_sql(query_reviews_corrigida, con=engine)
resultado_reviews_corrigida


avg_reviews_per_user
0                 24.33


<div class="alert alert-success">
<strong>Comentário do revisor v2</strong>

Valeu pelo ajuste!
</div>

### Resultado

Entre os usuários que avaliaram mais de 50 livros, a média de reviews escritas (com texto) foi de **24,33**.


<div class="alert alert-danger">
<strong>Comentário do revisor - v1</strong>

Quase isso!
    
Aqui precisamos usar duas tabelas: `ratings` e `reviews`.
    
Primeiro precisamos identificar os usuários que classificaram mais de 50 livros, usando a tabela `ratings`.
    
Depois, **a partir desses usuários**, precisamos **contar** o número médio de avaliações, usando a tabela `reviews`. Ou seja, devemos refazer essa última query considerando apenas os usuários filtrados na anterior.
    
Com essa dica, acredito que seja possível para você escrever uma _query_ que irá atender perfeitamente à solicitação do projeto.

</div>

<div class="alert alert-danger">
<strong>Comentário do revisor - v2</strong>

Ainda não é bem isso. O resultado final deve ficar entre 20 e 30.

O select dentro dos parênteses seleciona com sucesso os usuários com mais de 50 avaliações.
O desafio agora é contar quantas avaliações cada um desses usuários teve e depois tirar a média dessas contagens.

</div>

<div class="alert alert-danger">
<strong>Comentário do revisor - v3</strong>

Valencio, ainda não é bem isso. Vou compartilhar contigo a query aqui. Veja se ela faz sentido para você:

    """
    WITH active_users AS (
        SELECT
            username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50
    )
    SELECT
        ROUND(AVG(review_count), 2) AS avg_reviews_per_user
    FROM (
        SELECT
            username,
            COUNT(review_id) AS review_count
        FROM reviews
        WHERE username IN (SELECT username FROM active_users)
        GROUP BY username
    ) AS user_reviews_count;
    """

</div>

# Conclusão Geral – Análise SQL

A análise SQL permitiu extrair insights estratégicos a partir de dados de livros, autores, editoras, avaliações e reviews de usuários. O objetivo foi avaliar a composição do catálogo e o comportamento dos leitores da plataforma digital, com foco na preparação para o lançamento de um novo produto de leitura. Abaixo, os principais aprendizados extraídos:

1. **Atualidade do acervo**  
   Foram identificados **819 livros publicados após o ano 2000**, evidenciando que a plataforma mantém um catálogo moderno e relevante.

2. **Livros com maior engajamento**  
   Títulos como *Twilight*, *The Hobbit* e *Harry Potter* concentram o maior número de avaliações, revelando o apelo contínuo de obras de grande alcance editorial.

3. **Editoras com produção editorial densa**  
   A **Penguin Books** lidera com **42 livros com mais de 50 páginas**, indicando forte contribuição para o acervo de leitura prolongada — importante para curadoria e experiência de leitura contínua.

4. **Autores com alta reputação média**  
   A parceria **J.K. Rowling / Mary GrandPré** aparece com a maior média de classificação (4.28) entre livros com mais de 50 avaliações — sinal de excelência editorial com base estatisticamente significativa.

5. **Comportamento de superusuários**  
   Usuários que avaliaram mais de 50 livros escreveram, em média, **24,33 reviews com texto**. Esse grupo representa o público mais engajado da plataforma — produtores ativos de conteúdo, altamente valiosos para recomendação, testes de funcionalidades e construção de comunidades.

---

### Conclusão Final
O conjunto de dados revela uma plataforma com acervo atual, autores influentes, editoras relevantes e uma base de usuários participativa e ativa. Essas evidências oferecem um alicerce sólido para decisões estratégicas de produto — desde curadoria editorial até engajamento de superusuários — no lançamento de um novo app de leitura.

